actions/parser.go

```diff
@@ -11,74 +11,191 @@ import (
 	"net/http"
 )
 
-func uploadAndConvertClaimFile(r *http.Request) map[string]interface{} {
-	_ = r.ParseMultipartForm(ParseLowerBound << ParseUpperBound)
+func writeResponse(w http.ResponseWriter, response string) {
+	_, err := w.Write([]byte(response))
+	if err != nil {
+		fmt.Println(err)
+	}
+	fmt.Println(response)
+}
+
+func readClaimFile(w http.ResponseWriter, r *http.Request) []byte {
+	err := r.ParseMultipartForm(ParseLowerBound << ParseUpperBound)
+	if err != nil {
+		writeResponse(w, err.Error())
+		return nil
+	}
 
 	claimFile, _, err := r.FormFile(ClaimFileInputName)
 	if err != nil {
-		fmt.Println(err)
+		writeResponse(w, err.Error())
+		return nil
 	}
 	defer claimFile.Close()
 
 	claimFileBytes, err := io.ReadAll(claimFile)
 	if err != nil {
-		fmt.Println(err)
+		writeResponse(w, err.Error())
+		return nil
 	}
+
+	return claimFileBytes
+}
+
+func uploadAndConvertClaimFile(w http.ResponseWriter, r *http.Request) map[string]interface{} {
+	claimFileBytes := readClaimFile(w, r)
+	if claimFileBytes == nil {
+		return nil
+	}
+
 	var claimFileMap map[string]interface{}
-	err = json.Unmarshal(claimFileBytes, &claimFileMap)
+	err := json.Unmarshal(claimFileBytes, &claimFileMap)
 	if err != nil {
-		fmt.Println(err)
+		writeResponse(w, err.Error())
+		return nil
+	}
+
+	_, keyExists := claimFileMap[ClaimTag]
+	if !keyExists {
+		writeResponse(w, err.Error())
+		return nil
 	}
 	return claimFileMap[ClaimTag].(map[string]interface{})
 }
 
-func insertToClaimTable(r *http.Request, db *sql.DB, claimFileMap map[string]interface{}) {
-	versions := claimFileMap[VersionsTag].(map[string]interface{})
+func validateClaimKeys(claimFileMap map[string]interface{}) map[string]interface{} {
+	versions, keyExists := claimFileMap[VersionsTag].(map[string]interface{})
+	if !keyExists {
+		return nil
+	}
+
+	_, keyExists = versions["ocp"]
+	if !keyExists {
+		return nil
+	}
+
+	return versions
+}
+
+func insertToClaimTable(w http.ResponseWriter, r *http.Request, tx *sql.Tx, claimFileMap map[string]interface{}) bool {
+	versions := validateClaimKeys(claimFileMap)
 
 	// saving users input referring to who created claim file and partner's name
 	createdBy := r.FormValue(CreatedByInputName)
 	partnerName := r.FormValue(PartnerNameInputName)
 
-	_, err := db.Exec(InsertToClaimSQLCmd, versions["ocp"].(string),
-		createdBy, time.Now(), partnerName)
+	// missing fields in claim file or created_by field is null
+	if versions == nil || createdBy == "" {
+		writeResponse(w, MalformedClaimFileErr)
+		return false
+	}
+
+	_, err := tx.Exec(InsertToClaimSQLCmd, versions["ocp"].(string), createdBy, time.Now(), partnerName)
```

Comment:
Where have you started the transaction? Could you please point this out?

Reply:
at ParserHandler function, line 184

Reply:
I will look over it again